chore: upgrade typedoc

boneskull opened the pull request:
the way entry points are declared has changed; every package now needs its own `typedoc.json` (ugh).  both `@appium/base-driver` and `@appium/types` also need to ship their `typedoc.json` files, since they are consumed by external drivers.

Closes #18677.

boneskull committed:
chore: update typedoc entry point configs

the way entry points are declared has changed; every package now needs its own `typedoc.json` (ugh).  both `@appium/base-driver` and `@appium/types` also need to ship their `typedoc.json` files, since they are consumed by external drivers.

boneskull committed:
fix(docutils): fix display of output path

boneskull committed:
chore(typedoc-plugin-appium,docutils)!: upgrade typedoc

BREAKING CHANGE: `@appium/docutils` and `@appium/typedoc-plugin-appium` now require TypeDoc v0.24.x,TS v5.0.x, and `typedoc-plugin-markdown` v3.15.x.

Since v0.24.x had breaking changes, this addresses them.  It also eliminates the need for `typedoc-plugin-resolve-crossmodule-references`.  TypeDoc doesn't yet support TS v5.1.x.

boneskull commented:
The tests are broken and the build will fail

jlipps reviewed (approved):
code changes LGTM once tests are updated and commented code removed

boneskull committed:
chore(appium,base-driver,types): export more types

This fixes a whole lot of "type was referenced but not included in the documentation" warnings out of TypeDoc.  There are still more in the other packages, however.

- `lib/index.ts` in `appium` which re-exports stuff from `main.js` as well as a bunch of types.
- fixes some loose types in base-driver's `errors` module
- update ts peer dep

boneskull committed:
chore: fix most tests

boneskull committed:
fix(appium): correct type entry point

boneskull committed:
chore(appium): fix docs build

Because we need the root `typedoc.json` and we're building docs from `appium` (which now has its _own_ `typedoc.json`) we need to be explicit about where the `typedoc.json` / `tsconfig.json` actually lives.

boneskull committed:
fix(docutils): address mike/mkdocs PATH shenanigans; make fast typedoc.json parser

boneskull committed:
fix(docutils): fix --verbose

boneskull committed:
chore(docutils): add explicit type to avoid "any"

boneskull committed:
fix(docutils): use new bootstrap method and actually fail on error

boneskull committed:
fix(docutils): upgrade typescript

boneskull commented on packages/appium/docs/scripts/build-docs.js:
@jlipps take notice here: we need to be specific about paths now when running `appium-docs` programatically.

jlipps replied:
👍

boneskull commented on packages/appium/lib/main.js:
these are just missing exports

boneskull commented on packages/base-driver/lib/protocol/errors.js:
the types for all error classes are now exported from base-driver.  otherwise, this narrows some types.  wish lodash was more typesafe!

boneskull commented on packages/docutils/lib/builder/deploy.ts:
```diff
@@ -129,6 +129,17 @@ export async function deploy({
+  if (!mkdocsPath) {
+    throw new DocutilsError(`Could not find mkdocs executable; please run "${NAME_BIN} init`);
+  }
+  // okay, so if mkdocs is not in the PATH, then mike can't find it and will fail.
+  // so if we've found mkdocs, then we should add it to the path, just in case.
+  // further validation would involve asserting that the mkdocs version is what we expect,
+  // because this mkdocs executable may not be the same one we found via validation.
+  execOpts = _.defaultsDeep(execOpts, {
+    env: {PATH: `${path.dirname(mkdocsPath)}${path.delimiter}${process.env.PATH}`},
+  });
```
💀

jlipps replied:
lol

boneskull commented on packages/docutils/lib/fs.ts:
```diff
@@ -165,12 +166,12 @@ export const readPackageJson = _.memoize(_readPkgJson);
-  app.options.addReader(new TypeDocReader());
-  app.bootstrap({options: path.dirname(typedocJsonPath)});
-  return app.options.getRawValues();
+  // this is a null logger
+  const logger = new Logger();
+  const opts = new Options(logger);
+  // this mutates `opts`
+  new TypeDocReader().read(opts, logger, path.dirname(typedocJsonPath));
+  return opts.getRawValues();
```
the perf increase here is probably negligible, but this code is more direct anyway.

boneskull commented on packages/typedoc-plugin-appium/lib/converter/base-converter.ts:
here, and in many other files in the plugin, we previously had a `Context` object to work with, which contained the `ProjectDeclaration` reference we needed.  now, we only have the `ProjectDeclaration`.

boneskull commented on packages/typedoc-plugin-appium/lib/converter/builder.ts:
```diff
@@ -24,32 +30,23 @@ import {NAME_BUILTIN_COMMAND_MODULE} from './builtin-method-map';
-  ctx.postReflectionCreation(commandRefl, undefined, undefined);
-  ctx.finalizeDeclarationReflection(commandRefl);
+  project.registerReflection(commandRefl);
+  parent.children = [...(parent.children ?? []), commandRefl];
```
this is dubious, but I didn't see any other way I was supposed to do it.

boneskull committed:
chore(typedoc-plugin-appium): rename vars

boneskull commented on packages/typedoc-plugin-appium/lib/converter/utils.ts:
```diff
@@ -247,13 +247,10 @@ export function cloneTypeParameterReflection(
-    parentRefl,
-    tPRefl.varianceModifier
-  );
+  const newTPRefl = new TypeParameterReflection(tPRefl.name, parentRefl, tPRefl.varianceModifier);
+  newTPRefl.default = tPRefl.default;
+  newTPRefl.type = tPRefl.type;
+  return newTPRefl;
```
the constructor broke.

boneskull commented on packages/typedoc-plugin-appium/lib/options/reader.ts:
```diff
@@ -104,7 +109,19 @@ export class AppiumPluginOptionsReader implements OptionsReader {
 
     for (const reqdEntryPoint of REQUIRED_PACKAGES) {
-      const foundReqdEP = entryPoints.find((entryPoint) => entryPoint.includes(reqdEntryPoint));
+      const foundReqdEP = entryPoints.find((entryPoint) => {
+        if (hasMagic(entryPoint)) {
+          // this makes it find only dirs
+          entryPoint = entryPoint.replace(/\*$/, '*/');
+          try {
+            const matches = globSync(entryPoint);
+            return matches.includes(reqdEntryPoint);
+          } catch {
+            return false;
+          }
+        }
+        return entryPoint.includes(reqdEntryPoint);
+      });
```
now one can use globs for entry points in `typedoc.json`.  we need to check if any of those entry points match the required libs (`@appium/base-driver` and `@appium/types`).

boneskull commented:
TODO

- [ ] re-enable commented-out tests
- [ ] I think I might have broken package titles. fix

boneskull committed:
chore(types): remove checkJs from tsconfig

(there is no js in this workspace)

boneskull committed:
chore(typedoc-plugin-appium): more fixes

boneskull commented:
Current dependencies on/for this PR:
* master
  * **PR #18851** <a href="https://app.graphite.dev/github/pr/appium/appium/18851" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/appium/appium/18851?utm_source=stack-comment).

boneskull commented:
I'm still working on this when I get a min or two.  other than the package title thing being (probably) broken, it's just like a test fixture issue atm

jlipps commented:
nice!
